File: apps/portal/src/actions/usuarios.ts
'use server'

import { z } from 'zod'
import bcrypt from 'bcryptjs'
import { prisma } from '@/lib/prisma'
import { exigirAdmin } from '@/lib/auth'
import { revalidarTelas } from '@/lib/revalidar'
import { CARGO, MODULO } from '@/lib/dominio/cargos'

export type Resultado<T = void> = { ok: true; dados: T } | { ok: false; erro: string }

const cargos = Object.values(CARGO) as [string, ...string[]]
const modulos = Object.values(MODULO) as [string, ...string[]]

const opcional = z.string().trim().optional().transform((v) => (v ? v : undefined))

const esquemaCriacao = z.object({
  nome: z.string().trim().min(3, 'Informe o nome completo.'),
  email: z.string().trim().email('E-mail inválido.'),
  senha: z.string().min(8, 'A senha precisa de pelo menos 8 caracteres.'),
  cargo: z.enum(cargos),
  telefone: opcional,
  observacao: opcional,
  modulos: z.array(z.enum(modulos)).default([]),
})

export async function criarUsuario(entrada: unknown): Promise<Resultado<{ id: string }>> {
  await exigirAdmin()

  const parsed = esquemaCriacao.safeParse(entrada)
  if (!parsed.success) {
    return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }
  }
  const d = parsed.data

  try {
    const usuario = await prisma.usuario.create({
      data: {
        nome: d.nome,
        email: d.email.toLowerCase(),
        senhaHash: bcrypt.hashSync(d.senha, 10),
        cargo: d.cargo,
        telefone: d.telefone ?? null,
        observacao: d.observacao ?? null,
        acessos: { create: d.modulos.map((modulo) => ({ modulo })) },
      },
    })
    revalidarTelas('/usuarios')
    return { ok: true, dados: { id: usuario.id } }
  } catch (e) {
    const msg = e instanceof Error && e.message.includes('Unique')
      ? 'Já existe um usuário com esse e-mail.'
      : 'Falha ao cadastrar o usuário.'
    return { ok: false, erro: msg }
  }
}

const esquemaEdicao = z.object({
  cargo: z.enum(cargos),
  ativo: z.coerce.boolean(),
  modulos: z.array(z.enum(modulos)).default([]),
})

export async function editarUsuario(id: string, entrada: unknown): Promise<Resultado> {
  const admin = await exigirAdmin()

  // Ninguém tira o próprio acesso. Sem esta trava, um administrador distraído se rebaixa
  // para Consulta e não existe mais quem consiga desfazer — a recuperação exigiria mexer
  // no banco à mão.
  if (id === admin.id) {
    return { ok: false, erro: 'Você não pode mudar o próprio cargo nem se desativar. Peça a outro administrador.' }
  }

  const parsed = esquemaEdicao.safeParse(entrada)
  if (!parsed.success) {
    return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }
  }
  const d = parsed.data

  try {
    // Apaga e recria os acessos numa transação só: se apagar der certo e criar falhar, a
    // pessoa ficaria sem módulo nenhum sem ninguém ter pedido isso.
    await prisma.$transaction([
      prisma.usuario.update({ where: { id }, data: { cargo: d.cargo, ativo: d.ativo } }),
      prisma.acessoModulo.deleteMany({ where: { usuarioId: id } }),
      prisma.acessoModulo.createMany({ data: d.modulos.map((modulo) => ({ usuarioId: id, modulo })) }),
    ])
    revalidarTelas('/usuarios')
    return { ok: true, dados: undefined }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao salvar o usuário.' }
  }
}

const esquemaSenha = z.object({
  senha: z.string().min(8, 'A senha precisa de pelo menos 8 caracteres.'),
})

export async function redefinirSenha(id: string, entrada: unknown): Promise<Resultado> {
  await exigirAdmin()

  const parsed = esquemaSenha.safeParse(entrada)
  if (!parsed.success) {
    return { ok: false, erro: parsed.error.issues.map((i) => i.message).join(' ') }
  }

  try {
    await prisma.usuario.update({
      where: { id },
      data: { senhaHash: bcrypt.hashSync(parsed.data.senha, 10) },
    })
    revalidarTelas('/usuarios')
    return { ok: true, dados: undefined }
  } catch (e) {
    return { ok: false, erro: e instanceof Error ? e.message : 'Falha ao redefinir a senha.' }
  }
}

File: migracao-go/internal/application/identidade/usuarios.go
package identidade

import (
	"context"
	"errors"
	"net/mail"
	"strings"

	"golang.org/x/crypto/bcrypt"

	dominio "siqueiracampos/servidor/internal/domain/identidade"
)

// ErroValidacao junta as mensagens como o zod fazia em Next.js — ver actions/usuarios.ts.
type ErroValidacao struct{ Mensagens []string }

func (e *ErroValidacao) Error() string { return strings.Join(e.Mensagens, " ") }

func erroValidacao(msgs ...string) *ErroValidacao { return &ErroValidacao{Mensagens: msgs} }

type GerenciadorUsuarios struct {
	Usuarios dominio.UsuarioRepositorio
}

type EntradaCriarUsuario struct {
	Nome, Email, Senha, Cargo, Telefone, Observacao string
	Modulos                                         []string
	PapelFinanceiro                                 string
}

func ponteiroSeNaoVazio(s string) *string {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}
	return &s
}

func validarCargoEModulos(cargoStr string, modulosStr []string) (dominio.Cargo, []dominio.Modulo, []string) {
	var erros []string

	cargo := dominio.Cargo(cargoStr)
	if !cargo.Valido() {
		erros = append(erros, "Cargo inválido.")
	}

	modulos := make([]dominio.Modulo, 0, len(modulosStr))
	for _, m := range modulosStr {
		mm := dominio.Modulo(m)
		if !mm.Valido() {
			erros = append(erros, "Módulo inválido.")
			continue
		}
		modulos = append(modulos, mm)
	}
	return cargo, modulos, erros
}

func validarPapelFinanceiro(papelStr string) (dominio.PapelFinanceiro, []string) {
	papel := dominio.PapelFinanceiro(papelStr)
	if papel == "" {
		papel = dominio.PapelFinanceiroNenhum
	}
	if !papel.Valido() {
		return papel, []string{"Papel do Financeiro inválido."}
	}
	return papel, nil
}

// Criar espelha `criarUsuario` de actions/usuarios.ts — ver COMPORTAMENTO.md §5.
func (g *GerenciadorUsuarios) Criar(ctx context.Context, e EntradaCriarUsuario) (*dominio.Usuario, error) {
	var erros []string

	nome := strings.TrimSpace(e.Nome)
	if len(nome) < 3 {
		erros = append(erros, "Informe o nome completo.")
	}

	email := strings.ToLower(strings.TrimSpace(e.Email))
	if _, err := mail.ParseAddress(email); err != nil {
		erros = append(erros, "E-mail inválido.")
	}

	if len(e.Senha) < 8 {
		erros = append(erros, "A senha precisa de pelo menos 8 caracteres.")
	}

	cargo, modulos, errosCargo := validarCargoEModulos(e.Cargo, e.Modulos)
	erros = append(erros, errosCargo...)

	papelFinanceiro, errosPapel := validarPapelFinanceiro(e.PapelFinanceiro)
	erros = append(erros, errosPapel...)

	if len(erros) > 0 {
		return nil, erroValidacao(erros...)
	}

	hash, err := bcrypt.GenerateFromPassword([]byte(e.Senha), 10)
	if err != nil {
		return nil, err
	}

	u := &dominio.Usuario{
		Nome: nome, Email: email, SenhaHash: string(hash), Cargo: cargo, Ativo: true,
		Telefone: ponteiroSeNaoVazio(e.Telefone), Observacao: ponteiroSeNaoVazio(e.Observacao),
		PapelFinanceiro: papelFinanceiro,
	}

	if err := g.Usuarios.Criar(ctx, u, modulos); err != nil {
		if errors.Is(err, dominio.ErrEmailDuplicado) {
			return nil, erroValidacao("Já existe um usuário com esse e-mail.")
		}
		return nil, err
	}
	u.Modulos = modulos
	return u, nil
}

type EntradaEditarUsuario struct {
	Cargo           string
	Ativo           bool
	Modulos         []string
	PapelFinanceiro string
}

// Editar bloqueia autoedição — ver COMPORTAMENTO.md §5: sem essa trava um admin
// distraído se rebaixa e ninguém mais consegue desfazer sem mexer direto no banco.
func (g *GerenciadorUsuarios) Editar(ctx context.Context, admin dominio.Sessao, alvoID string, e EntradaEditarUsuario) error {
	if alvoID == admin.ID {
		return erroValidacao("Você não pode mudar o próprio cargo nem se desativar. Peça a outro administrador.")
	}

	cargo, modulos, erros := validarCargoEModulos(e.Cargo, e.Modulos)
	papelFinanceiro, errosPapel := validarPapelFinanceiro(e.PapelFinanceiro)
	erros = append(erros, errosPapel...)
	if len(erros) > 0 {
		return erroValidacao(erros...)
	}

	return g.Usuarios.AtualizarCargoEAcessos(ctx, alvoID, cargo, e.Ativo, modulos, papelFinanceiro)
}

func (g *GerenciadorUsuarios) RedefinirSenha(ctx context.Context, alvoID, novaSenha string) error {
	if len(novaSenha) < 8 {
		return erroValidacao("A senha precisa de pelo menos 8 caracteres.")
	}
	hash, err := bcrypt.GenerateFromPassword([]byte(novaSenha), 10)
	if err != nil {
		return err
	}
	return g.Usuarios.AtualizarSenha(ctx, alvoID, string(hash))
}
